feat: Add collab liquidation

holzeis opened the pull request:
Analog to the expired positions this is adding an optimistic approach to close liquidated positions collaboratively.

If the ask or the bid price gets within a range of 10% to the liquidation price, an async order with order reason `Liquidated` is automatically created and the user notified about the liquidated positions. The user will get daily reminders for a week to collab close the liquidated position.

This means with an average entry price of $30,000 the user gets liquidated on a short position at $37,500 - $3,750 = $33,750.


https://github.com/get10101/10101/assets/382048/0ab2da62-ffdf-415c-8382-1fe1ddba4171



fixes https://github.com/get10101/meta/issues/407

holzeis commented:
Should we show the margin call in the app?

holzeis commented:
The order reason is shown in the order history if the order reason is different to "manual"

![image](https://github.com/get10101/10101/assets/382048/d9e6f417-2d22-4b4e-93a5-e8e34317397d)

bonomat reviewed (changes requested):
I'm afraid I put you onto the wrong track here. The way how this is used and called at the moment is not a `margin call` and I also put you onto the wrong track on how it is calculated.
I'm sorry but I would rather not merge this as is then having a wrong terminology introduced which might confuse us and traders down the road.

Here is what we should do: 

Implement a maintenance margin and derive the liquidation price from this maintenance margin. 
I would then simply replace the current `Liquidation Price` with the calculation from below and optionally add a `Bankruptcy Price`: this is where the user loses everything. 

Here a full example: 

### Assumptions:
- Initial price of Bitcoin (BTC): $10,000
- Contract size: $10,000
- Initial margin requirement: 50% (leverage of 2x)
- Maintenance margin requirement: 10% (this is quite high but maybe OK for us)

### Scenario 1: Opening a Long Position
1. **Opening a Long Position**:
    - Contract size: $10,000 
    - Initial Leverage: 2x (you need to deposit 50% of the total contract size as the initial margin).
    - Initial Margin = 50% * $10,000 = $5,000
    - The initial margin in BTC using the current price of Bitcoin:
      BTC Initial Margin = $5,000 / $10,000 per BTC = 0.5 BTC

2. **Liquidation Price Calculation**:
    - This is the formula to calculate the liquidation price using maintenance margin: `(price_opening * leverage) / (leverage + 1 - (maintenance_margin * leverage))`
    - If maintenance_margin = 0%: 
      - Liquidation Price = (price_opening * leverage) / (leverage + 1 - (maintenance_margin * leverage))
        Liquidation Price = ($10,000 * 2) / (2 + 1 - (0.1 * 0))
        Liquidation Price = $20,000 / (3 - 0)
        Liquidation Price = $20,000 / 3
        Liquidation Price ≈ $6,666.67 
    - If maintenance_margin = 10%:
      - Liquidation Price = (price_opening * leverage) / (leverage + 1 - (maintenance_margin * leverage))
        Liquidation Price = ($10,000 * 2) / (2 + 1 - (0.1 * 2))
        Liquidation Price = $20,000 / (3 - 0.2)
        Liquidation Price = $20,000 / 2.8
        Liquidation Price ≈ $7,142.86
      - --> meaning, the user gets liquidated earlier

### Scenario 2: Opening a Short Position
1. **Opening a Short Position**:
    - Contract size: $10,000
    - Initial Leverage: 2x (you need to deposit 50% of the total contract size as the initial margin).
    - Initial Margin = 50% * $10,000 = $5,000
    - The initial margin in BTC using the current price of Bitcoin:
      BTC Initial Margin = $5,000 / $10,000 per BTC = 0.5 BTC

2. **Liquidation Price Calculation**:
    - If maintenance_margin = 0%:
      - Liquidation Price = (price_opening * leverage) / (leverage - 1 + (maintenance_margin * leverage))
        Liquidation Price = ($10,000 * 2) / (2 - 1 + (0.1 * 0))
        Liquidation Price = $20,000 / (2 - 1)
        Liquidation Price = $20,000 / 1
        Liquidation Price = $20,000
        Liquidation Price = $20,000 per BTC
   - If `maintenance_margin = 10%`:
       - Liquidation Price = (price_opening * leverage) / (leverage - 1 + (maintenance_margin * leverage))
         Liquidation Price = ($10,000 * 2) / (2 - 1 + (0.1 * 2))
         Liquidation Price = $20,000 / (2 - 1 + 0.2)
         Liquidation Price = $20,000 / (1 + 0.2)
         Liquidation Price = $20,000 / 1.2
         Liquidation Price ≈ $16,666.67
       - --> meaning, the user gets liquidated earlier

holzeis commented:
> I'm afraid I put you onto the wrong track here. The way how this is used and called at the moment is not a `margin call` and I also put you onto the wrong track on how it is calculated. I'm sorry but I would rather not merge this as is then having a wrong terminology introduced which might confuse us and traders down the road.
> 
> Here is what we should do:
> 
> Implement a maintenance margin and derive the liquidation price from this maintenance margin. I would then simply replace the current `Liquidation Price` with the calculation from below and optionally add a `Bankruptcy Price`: this is where the user loses everything.
> 
> Here a full example:
> ### Assumptions:
> 
>     * Initial price of Bitcoin (BTC): $10,000
> 
>     * Contract size: $10,000
> 
>     * Initial margin requirement: 50% (leverage of 2x)
> 
>     * Maintenance margin requirement: 10% (this is quite high but maybe OK for us)
> 
> 
> ### Scenario 1: Opening a Long Position
> 
>     1. **Opening a Long Position**:
>        
>        * Contract size: $10,000
>        * Initial Leverage: 2x (you need to deposit 50% of the total contract size as the initial margin).
>        * Initial Margin = 50% * $10,000 = $5,000
>        * The initial margin in BTC using the current price of Bitcoin:
>          BTC Initial Margin = $5,000 / $10,000 per BTC = 0.5 BTC
> 
>     2. **Liquidation Price Calculation**:
>        
>        * This is the formula to calculate the liquidation price using maintenance margin: `(price_opening * leverage) / (leverage + 1 - (maintenance_margin * leverage))`
>        * If maintenance_margin = 0%:
>          
>          * Liquidation Price = (price_opening * leverage) / (leverage + 1 - (maintenance_margin * leverage))
>            Liquidation Price = ($10,000 * 2) / (2 + 1 - (0.1 * 0))
>            Liquidation Price = $20,000 / (3 - 0)
>            Liquidation Price = $20,000 / 3
>            Liquidation Price ≈ $6,666.67
>        * If maintenance_margin = 10%:
>          
>          * Liquidation Price = (price_opening * leverage) / (leverage + 1 - (maintenance_margin * leverage))
>            Liquidation Price = ($10,000 * 2) / (2 + 1 - (0.1 * 2))
>            Liquidation Price = $20,000 / (3 - 0.2)
>            Liquidation Price = $20,000 / 2.8
>            Liquidation Price ≈ $7,142.86
>          * --> meaning, the user gets liquidated earlier
> 
> 
> ### Scenario 2: Opening a Short Position
> 
>     1. **Opening a Short Position**:
>        
>        * Contract size: $10,000
>        * Initial Leverage: 2x (you need to deposit 50% of the total contract size as the initial margin).
>        * Initial Margin = 50% * $10,000 = $5,000
>        * The initial margin in BTC using the current price of Bitcoin:
>          BTC Initial Margin = $5,000 / $10,000 per BTC = 0.5 BTC
> 
>     2. **Liquidation Price Calculation**:
>        
>        * If maintenance_margin = 0%:
>          
>          * Liquidation Price = (price_opening * leverage) / (leverage - 1 + (maintenance_margin * leverage))
>            Liquidation Price = ($10,000 * 2) / (2 - 1 + (0.1 * 0))
>            Liquidation Price = $20,000 / (2 - 1)
>            Liquidation Price = $20,000 / 1
>            Liquidation Price = $20,000
>            Liquidation Price = $20,000 per BTC
>        * If `maintenance_margin = 10%`:
>          
>          * Liquidation Price = (price_opening * leverage) / (leverage - 1 + (maintenance_margin * leverage))
>            Liquidation Price = ($10,000 * 2) / (2 - 1 + (0.1 * 2))
>            Liquidation Price = $20,000 / (2 - 1 + 0.2)
>            Liquidation Price = $20,000 / (1 + 0.2)
>            Liquidation Price = $20,000 / 1.2
>            Liquidation Price ≈ $16,666.67
>          * --> meaning, the user gets liquidated earlier

Thanks for your detailed explanation and education on the different liquidation prices :). I've addressed your remarks in 3d28628f2de182c6fd8d6f7ff06b3b04f855d124

Please have another look.

holzeis commented on coordinator/src/node/liquidated_positions.rs:
```diff
@@ -0,0 +1,533 @@
+
+/// Checks if the position needs to get liquidated. Either from the trader or the coordinator point
+/// of view.
+fn check_if_position_needs_to_get_liquidated(
```
> Also we shouldn't need to recalculate the liquidation_price inside the function. We should have this in the position right?

yes, this has been addressed here 46eaa476cf0732d855f355b5a37243d04ecb63a9

holzeis replied:
> 🔧 I find this function a bit too complex as it's checking both parties at the same time and returns a direction in which it's getting liquidated or None.
> 
> What's the reason for not having a function which only accepts one liquidation price?
> 
> It could look like this:
> 
> ```rust
> fn check_if_position_needs_to_get_liquidated(
>     trader_direction: Direction,
>     best_current_price: &Price,
>     liquidation_price: Decimal,
> ) -> bool {
> ...
> }
> ```
> 
> you can now call this twice, once for the coordinator and once for the trader.
> 
> The length of this function would then only be half what it's doing right now.

I refactored it quite significantly. Have a look here 0f1ca45f2371ae4711c0a4d97e0bd59e8ec92bf9

holzeis commented on coordinator/src/node/liquidated_positions.rs:
```diff
@@ -0,0 +1,533 @@
+        return Ok(());
+    }
+
+    let best_current_price = commons::best_current_price(&open_orders);
```
done 8675cc129dbbfeb2419742181cbd20d655628859

bonomat commented on crates/trade/src/cfd.rs:
```diff
@@ -46,18 +46,34 @@ pub fn calculate_quantity(opening_price: f32, margin: u64, leverage: f32) -> f32
+    calculate_long_liquidation_price(leverage, price, Decimal::ZERO)
+}
+
+pub fn calculate_long_liquidation_price(
```
Done myself.

holzeis replied:
thanks

bonomat commented on mobile/native/src/ln_dlc/mod.rs:
```diff
@@ -168,6 +169,13 @@ pub fn get_seed_phrase() -> Vec<String> {
+pub fn get_maintenance_margin() -> Decimal {
+    match state::try_get_tentenone_config() {
+        Some(config) => Decimal::try_from(config.maintenance_margin).expect("to fit into decimal"),
+        None => dec!(0.1),
```
Maybe good to log an error if this ever returns `None` or maybe even add an `expect` as it should always be set?
Otherwise this will just hide errors which are hard to find.

bonomat reviewed (approved):
Looks good. I've added a few comments which can be resolved later. I'll go ahead and rebase for you.

luckysori reviewed (commented):
Half-way through, but it got merged.

luckysori commented on coordinator/src/node/liquidated_positions.rs:
```diff
@@ -0,0 +1,533 @@
+            if let Some(bid) = best_current_price.bid {
+                // check if coordinator needs to get liquidated
+                let coordinator_liquidation_price =
+                    calculate_long_liquidation_price(coordinator_leverage, average_entry_price);
```
🔧 It's a bit more readable to calculate both liquidation prices at the top.

luckysori commented on coordinator/src/node/liquidated_positions.rs:
```diff
@@ -0,0 +1,533 @@
+}
+
+#[cfg(test)]
+mod tests {
```
🔧 We should avoid "Test-Driven Design Ossification"[^1].

In summary, we should not right tens of tests against the same API and instead call a data-driven test API that delegates to the API we want to test. I've used the `check` idiom [here](https://github.com/get10101/10101/pull/2352/files#diff-dc08eded476b7979ce9fa96345782e5d1c570f176643482d8baf65678b9e4477R947), for example.

[^1]: https://matklad.github.io/2021/05/31/how-to-test.html#Test-Driven-Design-Ossification.

luckysori replied:
This approach would have paid off twice before this PR even got merged lol.

holzeis replied:
The tests have been removed as after the refactoring they were basically only testing `>=` and `<=`.

luckysori commented on coordinator/src/scheduler.rs:
```diff
@@ -222,3 +245,42 @@ fn build_remind_to_close_expired_position_notification_job(
+) -> Result<Job, JobSchedulerError> {
+    Job::new_async(schedule, move |_, _| {
+        let notification_sender = notification_sender.clone();
+        let mut conn = pool.get().expect("To be able to get a db connection");
```
📣 Looks to me like a risky `expect`.

holzeis replied:
https://github.com/get10101/10101/pull/2371

luckysori commented on coordinator/migrations/2024-04-03-105900_add_coordinator_liquidation_price/up.sql:
```diff
@@ -0,0 +1,5 @@
+    ADD COLUMN coordinator_liquidation_price REAL NOT NULL DEFAULT 0;
+
+UPDATE positions SET coordinator_liquidation_price = average_entry_price * coordinator_leverage / (coordinator_leverage + 1) where trader_direction='short';
+UPDATE positions SET coordinator_liquidation_price = average_entry_price * coordinator_leverage / (coordinator_leverage - 1) where trader_direction='long';
```
Hmmm, let's fix this though ❗

bonomat commented on coordinator/migrations/2024-04-02-110300_add_liquidation/down.sql:
```diff
@@ -0,0 +1,2 @@
+-- type changes can not be reverted
+select 0 from trade_params;
```
We need at least one query here or diesel will fail. I most of the time do a simple `select 1;`

luckysori commented on coordinator/src/trade/mod.rs:
```diff
@@ -516,18 +517,37 @@ impl TradeExecutor {
     ) -> Result<()> {
-        let liquidation_price = liquidation_price(trade_params);
+        let price = trade_params.average_execution_price();
+        let maintenance_margin = { self.node.settings.read().await.maintenance_margin };
```
The name suggests that this is an `Amount`, but it's actually a percentage.

holzeis replied:
https://github.com/get10101/10101/pull/2372

holzeis commented on coordinator/src/node/liquidated_positions.rs:
```diff
@@ -0,0 +1,517 @@
+    use trade::Direction;
+
+    #[test]
+    fn test_no_liquidatation_of_users_short_position_before_maintenance_margin() {
```
yeah.. i thought about removing them, but they were already there so I guess they don't hurt 🤷‍♂️

holzeis commented on coordinator/src/notifications.rs:
```diff
@@ -120,6 +122,10 @@ fn build_notification(kind: &NotificationKind) -> fcm::Notification<'_> {
         }
+        NotificationKind::PositionLiquidated => {
+            notification_builder.title("Your position has been liquidated");
+            notification_builder.body("Close your position.");
```
We need the user to open the app to collaboratively execute the liquidation.

> Your position has been liquidated 📉😞 

Don't know if you'd have to be sad about it. It could also be that the counterparty got liquidated and you made a huge profit :)

> Open the app to try again!

I copied it from the expired position notification as its the same action the user needs to do, which is "open the app". I guess we should change it to this message :)

holzeis replied:
https://github.com/get10101/10101/pull/2374

luckysori replied:
> Don't know if you'd have to be sad about it. It could also be that the counterparty got liquidated and you made a huge profit :)

I guess the message should depend on who got liquidated!

holzeis commented on coordinator/src/node/liquidated_positions.rs:
```diff
@@ -0,0 +1,533 @@
+
+/// The timeout before we give up on closing a liquidated position collaboratively. This value
+/// should not be larger than our refund transaction time lock.
+pub const LIQUIDATION_POSITION_TIMEOUT: Duration = Duration::days(7);
```
Thanks for the reminder, this timeout only defines how long we try to collaborate with the user before we consider force closing the channel

holzeis commented on coordinator/src/node/liquidated_positions.rs:
```diff
@@ -0,0 +1,517 @@
+        let liquidation = check_if_position_needs_to_get_liquidated(
+            trader_direction,
+            &price,
+            // the liquidation price of the trader is at 30,000 * 5 / 4 = 37,500
```
yes it is

holzeis replied:
I removed the tests altogether since they aren't very valuable anymore.

https://github.com/get10101/10101/pull/2372

holzeis commented on coordinator/src/node/liquidated_positions.rs:
```diff
@@ -0,0 +1,533 @@
+        let trader_leverage =
+            Decimal::try_from(position.trader_leverage).expect("to fit into decimal");
+
+        if let Some(direction) = check_if_position_needs_to_get_liquidated(
```
changed in a later commit.

luckysori commented on coordinator/src/node/liquidated_positions.rs:
```diff
@@ -0,0 +1,533 @@
+fn check_if_position_needs_to_get_liquidated(
+    trader_direction: Direction,
+    average_entry_price: Decimal,
+    best_current_price: &Price,
```
This got simplified later, I think.
